Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +127,185 @@
+%    This could be inner-ly repeated PLDs for some implementation reason or because of multiTE. We check both and issue a warning if more PLDs than TEs are present, as we don't really implement this option
+%    properly now. 
+[~,~,indexFirstPLD] = unique(xQ.Initial_PLD); % Find locations of all PLDs
+indexSecondPLD = find(indexFirstPLD==2); % Find the location of the second PLD
+if isempty(indexSecondPLT)
+	nInnerRepetitions = length(xQ.Initial_PLD); % There's no second PLD and nInnerRepetitions is the length of the PLD vector
+else
+	nInnerRepetitions = indexSecondPLD - 1; % The number of first-PLDs, i.e. the count of all PLDs before the second one, is the number of inner repetitions
+end
 
-%% 2. Reorder multi-TE data
+if nInnerRepetitions ~= xQ.NumberEchoTimes % Check for TE number differ from the number of inner repeats as this is a special option needing a special reordering
+	warning('Number of inner repetitions and echo-times does not match for multi-TE Hadamard. This option is not yet properly implemented.');
+end
+
+% 4. We define how many volumes are within an inner repetition accounting for the number of blocks within a Hadamard volume
+nVolumesPerInnerRepetition = xQ.TimeEncodedMatrixSize * nInnerRepetitions;
+
+% 5. We calculate the number of full repetitions - that is reapeating the entire acquisition scheme
+nFullRepetitions = nEncodedVolumes / nVolumesPerInnerRepetition; 
+
+
+%% 4. Obtain control4D images
+% Here we select all TEs for the control images
+% For example for 64 volumes and 2 repetitions with 8 PLDs and 4 TEs, it takes volume 1,2,3,4 and 33,34,35,36 to get all TEs of the control
+indexControl = ((1:nInnerRepetitions)'*ones(1,nFullRepetitions) + ones(nInnerRepetitions,1)*(0:nFullRepetitions-1)*nVolumesPerInnerRepetition);
+indexControl = indexControl(:)';
+
+decodedControl4D = imEncoded(:,:,:, indexControl);
+xQ.EchoTime_Control4D = xQ.EchoTime(indexControl);
+xQ.InitialPLD_Control4D = xQ.Initial_PLD(indexControl);
+xQ.LabelingDuration_Control4D = xQ.LabelingDuration(indexControl);
+
+%% 5. Reorder image matrix
+% Reorders that data for innerRepetitions. The example below shows it for the case of innerRepetitions being TEs only:
 % At this point the data is organized like this (in terms of ASL4D.nii volumes):
 % PLD1/TE1,PLD1/TE2,PLD1/TE3,PLD1/TE4...PLD2/TE1,PLD2/TE2,PLD2/TE3... (TEs in first dimension, PLDs after)
 %
-% And for decoding we want
+% And for decoding we want to swap the dimensions for TE and PLD so that:
 % TE1/PLD1,TE1/PLD2,TE1/PLD3,TE1/PLD4...TE2/PLD1,TE2/PLD2,TE2/PLD3,TE2/PLD4 (PLDs in the first dimension, TEs after)
-if isfield(xQ,'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
+% Then below, we apply this "transformation"/"decoding" vector to the image matrix and to the parameters TE/PLD/LD
+if nInnerRepetitions > 1)
```

Comment:
We always switch there and back. But it's better to have this. As for innerrepetitions == 1, this will do unnecessary data manipulation. So it is much cleaner for the user to show him that we really don't do anything. So that the user isn't confused what is hapenning with the data with no  inner repetitions.